### DISCUSSION TITLE
Generate linter friendly output by default

### DISCUSSION THREAD

**Discussion by abitrolly** [sqlalchemy/alembic#925] [Usage Questions]
**Describe the bug**

Generated migrations are not linter friendly. `black` fails, because migrations use single quotes. [`isort`](https://github.com/PyCQA/isort) because imports are not properly ordered.
```diff
-from alembic import op
-import sqlalchemy as sa
+import sqlalchemy as sa
+
+from alembic import op
```
 Perhaps there are more issues, but at least these two could already make it better.

**Expected behavior**

Generated migration to be `black` and `isort` compatible.

**Comment by CaselIT** [comment]
Hi,

you are free to modify the template file however you want.
Regarding the formatting, anyone has its own preference, so to address that alembic supports ["post_write_hooks"](https://alembic.sqlalchemy.org/en/latest/autogenerate.html#post-write-hooks) for exactly this use case.

**Reply by abitrolly** [reply]
Yes, I can modify template and install web hooks, but that's doesn't answer the question why not at least try to make the `alembic` output linter friendly by default?

**Reply by CaselIT** [reply]
For my part, the main reason is that the ratio usefulness / time spent implementing it is very low for such a feature, considering that are tools specifically built for such purpose, and that alembic does not always has the control on how things are rendered, using `repr` in multiple places.
Also, since formatting is a highly personal thing, it's impossible to please everybody, so we would get issues like "I don't like the tailing comma" or "I prefer single quotes" that are best addressed by a specific tool.

What we could do it maybe add an option to the `init` command to make it autogenerate post_write_hoocks for known configurations, something like `alembic init --post-write black isort`

**Reply by zzzeek** [reply]
@caselit we've implemented this already :)

**Reply by zzzeek** [reply]
oh, poster is complaining that's not good enough, OK

**Reply by zzzeek** [reply]
> Yes, I can modify template and install web hooks, but that's doesn't answer the question why not at least try to make the `alembic` output linter friendly by default?

this is not by "default" because it would make no sense for alembic to re-implement the code formatting job that black already does.   from there, black is not an installation dependency, as @CaselIT  mentions, there are other code formatters and also it's not a given that people are running autogenerate at all.   lots and lots of effort went into our code formatting integration so I would ask that you please use the feature as documented, thanks!

**Reply by abitrolly** [reply]
@zzzeek the poster probably needed to highlight the **default** from the start.

It took me about one hour to fix the formatting in https://github.com/pypa/warehouse/pull/9972 because tests are not being run immediately. If Alembic had a linter friendly output **by default** I would spent this hour doing something else. I could put a blame on PyPI maintainers for not putting the proper config in their repo, but I thought there might be a better way to fix the root cause.

**Reply by abitrolly** [reply]
@zzzeek poster just asked to replace single quote with double and reorder imports. Re-implementing black is not the goal to make the output friendly.

**Reply by zzzeek** [reply]
im not sure im understanding this.     you are using alembic, you (or your organization), when they went to configure alembic, chose not to use the write_hooks feature, and now the issue is that the thing that write_hooks is intended to do did not happen.

alembic requires configuration.  it's right up in the tutorial. are you arguing that alembic's requirement that it be configured is unacceptable?   i dont think that requirement is going to change, so what is your goal here?

**Reply by zzzeek** [reply]
in particular, "reordering imports" has *many* different approaches, there are many tools which do this, all differently.   it's not possible that alembic could choose an import reordering approach that would match whatever linters you're using, so what you ask is not actually even possible.

**Reply by abitrolly** [reply]
@zzzeek I am not using `alembic`, but the project on GitHub for which I submitted a PR is. I don't know why there is `write_hooks` feature. Maybe they considered that write hooks are insecure, or maybe they've started before this feature was implemented, or maybe the Pyramid framework considers running shell commands insecure. I don't know.

> alembic requires configuration. it's right up in the tutorial. are you arguing that alembic's requirement that it be configured is unacceptable?

No, I am not arguing. You say that `alembic` requires configuration, but I don't see one in the Pyramid, so I just proposed to change the non-critical default output.

> in particular, "reordering imports" has many different approaches

I doubt there is a linter that produces the output that `alembic` does. If there was one, I wouldn't touch this. 

> so what you ask is not actually even possible.

This is what I ask for https://github.com/sqlalchemy/alembic/pull/926

**Reply by zzzeek** [reply]
OK, your original issue was "Generated migrations are not linter friendly. "  - the generated migrations are a product of a whole python rendering thing which necessarily needs to run through a tool like black to be nicely formatted.

if OTOH you are merely looking *just for the two imports in the default template*, that's a much narrower scope.  but you mentioned "quotes" and things also.  we rely on the formatting tools for that stuff.

your PR is fine (though it would have to include all the templates, not just that single one) but it doesn't address the whole problem you're referencing, if generated migrations are to work with linters then one needs to configure the post_write hooks.  

I think your issue is still that "warehouse" is not configuring alembic completely.  Why cant you take this issue up with them?

**Reply by abitrolly** [reply]
> your PR is fine (though it would have to include all the templates, not just that single one) but it doesn't address the whole problem you're referencing, if generated migrations are to work with linters then one needs to configure the post_write hooks.

Yes, reordering imports doesn't fix the problem with single quotes, and as I don't know how the quotes are generated (manual AST walking or by a framework), my intention was just to fill an issue to see how far it can go. Neither I can test my changes right now.

**Reply by CaselIT** [reply]
> @zzzeek poster just asked to replace single quote with double and reorder imports. Re-implementing black is not the goal to make the output friendly.

The original request was a lot broader, but in any case my reply holds, since even replacing single quotes with double is not immediate, considering also that python renders string with single quotes in repr.

> If Alembic had a linter friendly output by default I would spent this hour doing something else. I could put a blame on PyPI maintainers for not putting the proper config in their repo, but I thought there might be a better way to fix the root cause.

Sorry but I think the issue lays with the pypi maintainers here. Alembic can be configured to respect a particular use formatting, so if a project requires one, it should be configured. There is "no root cause", since formatting is not an universal thing, and therefore cannot be fixed in an universal way. Alembic job is to render syntactically correct python, not to format it according the the particular formatting rule of a single project.

> It took me about one hour to fix the formatting in pypa/warehouse#9972 because tests are not being run immediately. If Alembic had a linter friendly output by default I would spent this hour doing something else.

I think the solution to this would be to improve `warehouse` contributors guidelines so that contributors can follow a simple workflow. Again "a linter friendly output by default" does not exists, since the output depends on the linter used.

**Reply by CaselIT** [reply]
> @CaselIT we've implemented this already :)

I don't think so. `alembic init --help` shows this:
```
usage: alembic init [-h] [-t TEMPLATE] [--package] directory

positional arguments:
  directory             location of scripts directory

optional arguments:
  -h, --help            show this help message and exit
  -t TEMPLATE, --template TEMPLATE
                        Setup template for use with 'init'
  --package             Write empty __init__.py files to the environment and version locations
```

**Comment by zzzeek** [comment]
hi there -

Alembic supports the feature you request fully.  Please see:   https://alembic.sqlalchemy.org/en/latest/autogenerate.html?highlight=black#applying-post-processing-and-python-code-formatters-to-generated-revisions  which will illustrate how to plug in the "black" code formatter, or any other formatters you'd like.

**Reply by abitrolly** [reply]
This is needed by Pyramid here https://github.com/pypa/warehouse which uses Alembic, but I don't see any `alembic.ini` files there.

**Reply by zzzeek** [reply]
i would advise to take the issue up with warehouse that their alembic configuration story needs to improve.

**Reply by abitrolly** [reply]
https://github.com/pypa/warehouse/issues/10053

**Comment by zzzeek** [comment]
there are many import ordering styles besides isort.  We use [zimports](https://pypi.org/project/zimports/) which will make use of any of the ordering styles used by [flake8-import-order](https://github.com/PyCQA/flake8-import-order), and we use the "google" style.  The imports as they are in the template right now are already compliant with "google" and changing them as proposed makes them non-compliant.   these templates are generated into the user-project directory, which is not the "alembic" project; therefore both the "sqlalchemy" and "alembic" imports are third party imports, and should be together and alphabetically ordered.

as stated previously, there is no fixed import order in Python so people need to use the post_write_hooks to install the formatting they prefer.

**Reply by zzzeek** [reply]
confirmed, "isort" and "zimports w google style" will sort the following file in two different ways

```py
from alembic import op
import sqlalchemy as sa

sa.foo
op.bar


```

**Reply by abitrolly** [reply]
Where is this "google" style spec you are referencing to? I can't validate your claim without seeing the actual spec.

**Reply by abitrolly** [reply]
@zzzeek the link to `zimports` in your post is broken too.

**Reply by CaselIT** [reply]
> 
> 
> @zzzeek the link to `zimports` in your post is broken too.

fixed

**Reply by CaselIT** [reply]
> 
> 
> Where is this "google" style spec you are referencing to? I can't validate your claim without seeing the actual spec.

https://github.com/PyCQA/flake8-import-order#styles

**Reply by abitrolly** [reply]
This is not what Google guidelines recommend.
```
import Z
from Z import A
from Z.A import A
from Z.A.B import A
```
They also do not contain any examples of `*` imports, which are generally discouraged.

**Reply by CaselIT** [reply]
From the link above: https://github.com/PyCQA/flake8-import-order/blob/master/tests/test_cases/complete_google.py

has an example of `*` import

**Reply by abitrolly** [reply]
Tthe official Google reference contains no such examples - https://google.github.io/styleguide/pyguide.html

**Reply by zzzeek** [reply]
https://google.github.io/styleguide/pyguide.html?showone=Imports_formatting#313-imports-formatting

> Within each grouping, imports should be sorted lexicographically, ignoring case, according to each module’s full package path (the path in from path import ...). Code may optionally place a blank line between import sections.

```
import collections
import queue
import sys

from absl import app
from absl import flags
import bs4
import cryptography
import tensorflow as tf

from book.genres import scifi
from myproject.backend import huxley
from myproject.backend.hgwells import time_machine
from myproject.backend.state_machine import main_loop
from otherproject.ai import body
from otherproject.ai import mind
from otherproject.ai import soul

# Older style code may have these imports down here instead:
#from myproject.backend.hgwells import time_machine
#from myproject.backend.state_machine import main_loop
```

**Reply by zzzeek** [reply]
openstack uses this style as well, which is why I'm used to it

https://docs.openstack.org/hacking/latest/user/hacking.html#imports

**Reply by abitrolly** [reply]
OpenStack explicitly denies wildcards.
```
* [H303] Do not use wildcard * import (*)
```
And also explicitly places third-party imports separated by blank line before projects imports. According to OpenStack definition of third-party modules the `import sqlalchemy as sa` is 3rd-party for Alembic.

> For the purposes of import order, OpenStack projects other than the one to which the file belongs are considered “third party”. Only imports from the same Git repo are considered “project imports”

**Reply by CaselIT** [reply]
we do follow that rule, see for example https://github.com/sqlalchemy/alembic/blob/master/alembic/operations/batch.py

**Reply by zzzeek** [reply]
I might have pointed this out earler.    The "sqlalchemy" and "alembic" imports as stated inside of a migration template **are both third party**, because this is a Python file that renders in the user project where alembic is a third party dependency.   that's why they are together in the current template.

**Reply by zzzeek** [reply]
note real world example in openstack:

https://github.com/openstack/neutron/blob/master/neutron/db/migration/alembic_migrations/dvr_init_opts.py#L18

**Reply by zzzeek** [reply]
their template: https://github.com/openstack/neutron/blob/master/neutron/db/migration/alembic_migrations/script.py.mako#L16

**Reply by abitrolly** [reply]
> I might have pointed this out earler. The "sqlalchemy" and "alembic" imports as stated inside of a migration template are both third party,

Yes. I didn't understand that, the issue is closed. `warehouse` project should implement their own config to call two external tools via shell when generating migrations to format the migration output.
